src/Spotify/SpotifyBackend.cpp

```diff
@@ -10,39 +10,279 @@
+    } else {                                                             \
+      tokenString = mSpotifyAuth.getAccessToken();                       \
+      returnValue = functionCall;                                        \
+                                                                         \
+      if (auto err = std::get_if<Error>(&returnValue)) {                 \
+        LOG(ERROR) << error->getErrorMessage() << std::endl;             \
+                                                                         \
+        return *err;                                                     \
+      }                                                                  \
+    }                                                                    \
+  }
+
+#define SPOTIFYCALL_WITH_REFRESH_OPT(returnValue, functionCall, tokenString) \
+  returnValue = functionCall;                                                \
+  if (returnValue.has_value()) {                                             \
+    auto ret = errorHandler(returnValue.value());                            \
+    if (ret.has_value()) {                                                   \
+      LOG(ERROR) << ret.value().getErrorMessage() << std::endl;              \
+      return ret.value();                                                    \
+    } else {                                                                 \
+      tokenString = mSpotifyAuth.getAccessToken();                           \
+      returnValue = functionCall;                                            \
+                                                                             \
+      if (returnValue.has_value()) {                                         \
+        LOG(ERROR) << returnValue.value().getErrorMessage() << std::endl;    \
+                                                                             \
+        return returnValue.value();                                          \
+      }                                                                      \
+    }                                                                        \
+  }
 
 TResultOpt SpotifyBackend::initBackend() {
-  return nullopt;
+  // start server for authentication
+  auto startServerRet = mSpotifyAuth.startServer();
+
+  return startServerRet;
 }
 
-TResult<vector<BaseTrack>> SpotifyBackend::queryTracks(string const &pattern,
-                                                       size_t const num) {
-  auto tmp = vector<BaseTrack>();
-  tmp.emplace_back();
-  return tmp;
+TResult<std::vector<BaseTrack>> SpotifyBackend::queryTracks(
+    std::string const &pattern, size_t const num) {
+  std::string token = mSpotifyAuth.getAccessToken();
+
+  TResult<SpotifyPaging> retVal;
+  SPOTIFYCALL_WITH_REFRESH(
+      retVal, mSpotifyAPI.search(token, pattern, QueryType::track, num), token);
+
+  auto page = std::get<SpotifyPaging>(retVal);
+  std::vector<BaseTrack> tracks;
+
+  for (auto const &elem : page.getTracks()) {
```

Comment:
in the case of a "random" pattern, `page.getTracks()` can potentially return an empty vector, do I see that correctly?

Reply:
jep, vector can be empty if nothing has been found